[llvm-objcopy][ELF] Add an option to remove notes

igorkudrin opened the pull request:
This adds an option `--remove-note` to selectively delete notes in an ELF file. For now, it is expected to be useful for editing core dump files; in particular, it searches for the notes in `PT_NOTE` segments and does not handle nested segments. The implementation can be extended later as needed.

RFC: https://discourse.llvm.org/t/rfc-llvm-objcopy-feature-for-editing-notes/83491

igorkudrin committed:
[llvm-objcopy][ELF] Add an option to remove notes

This adds an option `--remove-note` to selectively delete notes in an
ELF file. For now, it is expected to be useful for editing core dump
files; in particular, it searches for the notes in `PT_NOTE` segments
and does not handle nested segments. The implementation can be extended
later as needed.

jh7370 reviewed (commented):
Please update the Command Guide documentation under llvm/docs with the new option.

Just some initial thoughts as I don't have time to go through this in much depth yet.

I'm concerned that this approach seems to only work with notes in segments, when the option seems to apply reasonably to ET_REL objects too, where the notes are in sections only. Indeed, more generally llvm-objcopy is designed to work primarily with sections, with direct interactions with segments being very limited. I think it would more naturally fit with llvm-objcopy's design to work with note sections first. You could swap the section contents is the same way as other replace section operations work, so that much of the existing writing behaviour works unchanged.

Clearly, something special would need to be done for PT_NOTE segments, especially those without sections. One option might be to create an artificial section that covers the whole segment (or any parts of the segment not covered by an existing section at least) so that the section-based approach outlined above would still work, then have special handling to adjust the file/memsz fields after the sections have been updated.

Somewhat related to the previous point, I'm concerned with how `updateSegmentData` might interact with `--remove-section` for sections in segments. The current approach of the latter is that removed section data is overwritten with null bytes, preserving the segment size. I understand that for a PT_NOTE this may not be quite so desirable, but I think we need to do something to prevent mixing the two behaviours, because what the end result should be is non-obvious. One way would be to emit an error if a note section/segment is touched by both --remove-section and --remove-note.

igorkudrin commented on llvm/tools/llvm-objcopy/ObjcopyOptions.cpp:
```diff
@@ -527,6 +527,37 @@ static Expected<NewSymbolInfo> parseNewSymbolInfo(StringRef FlagValue) {
+    return createStringError(errc::invalid_argument,
+                             "bad format for --remove-note, missing type_id");
+  SmallVector<StringRef, 2> Tokens;
+  FlagValue.split(Tokens, '/', /*MaxSplit=*/1);
```
Since most parts of the argument are designed to be optional (see [RFC](https://discourse.llvm.org/t/rfc-llvm-objcopy-feature-for-editing-notes/83491)), the separators should be unique, otherwise, we would end up guessing whether a given token means the section or the originator's name. `:` is often used for sections, so I chose `/` for the note name. I don't have any strict preferences for the delimiters, though.

igorkudrin committed:
Merge branch 'main' into llvm-objcopy-remove-note

igorkudrin committed:
fixup: update parsing the argument

igorkudrin committed:
fixup: update tests

igorkudrin committed:
fixup: add the option to the command guide

igorkudrin commented:
> Please update the Command Guide documentation under llvm/docs with the new option.

Done

> Just some initial thoughts as I don't have time to go through this in much depth yet.
> 
> I'm concerned that this approach seems to only work with notes in segments, when the option seems to apply reasonably to ET_REL objects too, where the notes are in sections only. Indeed, more generally llvm-objcopy is designed to work primarily with sections, with direct interactions with segments being very limited. I think it would more naturally fit with llvm-objcopy's design to work with note sections first. You could swap the section contents is the same way as other replace section operations work, so that much of the existing writing behaviour works unchanged.
> 
> Clearly, something special would need to be done for PT_NOTE segments, especially those without sections. One option might be to create an artificial section that covers the whole segment (or any parts of the segment not covered by an existing section at least) so that the section-based approach outlined above would still work, then have special handling to adjust the file/memsz fields after the sections have been updated.

Thanks! I'll try to rework the patch using the sections-first approach.

> Somewhat related to the previous point, I'm concerned with how `updateSegmentData` might interact with `--remove-section` for sections in segments. The current approach of the latter is that removed section data is overwritten with null bytes, preserving the segment size. I understand that for a PT_NOTE this may not be quite so desirable, but I think we need to do something to prevent mixing the two behaviours, because what the end result should be is non-obvious. One way would be to emit an error if a note section/segment is touched by both --remove-section and --remove-note.

For the initial implementation, I plan to simply refuse to remove notes that reside in PT_LOAD segments. It seems that it is possible to actually remove data in non-load segments and update their contents and size accordingly. The sections in such segments and sections without segments (i.e. in ET_REL files) can also be updated in this way. Later, we can update the `--remove-section` command for non-load segments to work similarly.

igorkudrin committed:
fixup: fix updating offsets

igorkudrin committed:
fixup: split 'removeNote()'

igorkudrin committed:
fixup: do not allow '--remove-note' with '--(remove|add|update)-section'

igorkudrin committed:
fixup: change ELF type of the test file ET_CORE -> ET_EXEC

For core files, llvm-readobj dumps note segments as a whole,
without reporting individual sections.

igorkudrin committed:
fixup: also run the test for ELF32 and BE

igorkudrin committed:
fixup: handle SHT_NOTE sections outside of segments

igorkudrin commented:
> I'm concerned that this approach seems to only work with notes in segments, when the option seems to apply reasonably to ET_REL objects too, where the notes are in sections only. Indeed, more generally llvm-objcopy is designed to work primarily with sections, with direct interactions with segments being very limited. I think it would more naturally fit with llvm-objcopy's design to work with note sections first. You could swap the section contents is the same way as other replace section operations work, so that much of the existing writing behaviour works unchanged.

We can't rely on the existing code to update `SHT_NOTE` sections that reside in `PT_NOTE` segments. Currently, when the contents of such a section are replaced, the new data is written over the current segment contents, and if the section is shortened, the gap retains the old data. For `PT_NOTE` segments, this means that either some notes are preserved or duplicated (if the gap starts at the bound of a note), or the note entries in the segment starting from that point are corrupted. Thus, the offsets and sizes of such sections and the size of such segments should be adjusted.

However, the existing code is suitable for sections that are not associated with segments. I've updated the patch, so sections in relocatable objects are also supported now.

> Clearly, something special would need to be done for PT_NOTE segments, especially those without sections. One option might be to create an artificial section that covers the whole segment (or any parts of the segment not covered by an existing section at least) so that the section-based approach outlined above would still work, then have special handling to adjust the file/memsz fields after the sections have been updated.

This would either expose these temporary sections in the output file, or require some non-trivial changes to the code. And yet this does not solve the issue of gaps in `PT_NOTE` segments.

> Somewhat related to the previous point, I'm concerned with how `updateSegmentData` might interact with `--remove-section` for sections in segments. The current approach of the latter is that removed section data is overwritten with null bytes, preserving the segment size. I understand that for a PT_NOTE this may not be quite so desirable, but I think we need to do something to prevent mixing the two behaviours, because what the end result should be is non-obvious. One way would be to emit an error if a note section/segment is touched by both --remove-section and --remove-note.

I suspect that there may also be some problems with using the new command together with `--add-section` and `--update-section`, as the correct order of execution is not clear. I think the simplest solution is just to reject such combinations of options, so I've updated the patch accordingly.

jh7370 commented:
> > I'm concerned that this approach seems to only work with notes in segments, when the option seems to apply reasonably to ET_REL objects too, where the notes are in sections only. Indeed, more generally llvm-objcopy is designed to work primarily with sections, with direct interactions with segments being very limited. I think it would more naturally fit with llvm-objcopy's design to work with note sections first. You could swap the section contents is the same way as other replace section operations work, so that much of the existing writing behaviour works unchanged.
> 
> We can't rely on the existing code to update `SHT_NOTE` sections that reside in `PT_NOTE` segments. Currently, when the contents of such a section are replaced, the new data is written over the current segment contents, and if the section is shortened, the gap retains the old data. For `PT_NOTE` segments, this means that either some notes are preserved or duplicated (if the gap starts at the bound of a note), or the note entries in the segment starting from that point are corrupted. Thus, the offsets and sizes of such sections and the size of such segments should be adjusted.

Yeah, you'd need to compress the sections together and then shrink the segment. I'm not sure if there's existing code to modify a segment in this way yet.
 
> However, the existing code is suitable for sections that are not associated with segments. I've updated the patch, so sections in relocatable objects are also supported now.

I'm happy with segment support being added later.

> > Clearly, something special would need to be done for PT_NOTE segments, especially those without sections. One option might be to create an artificial section that covers the whole segment (or any parts of the segment not covered by an existing section at least) so that the section-based approach outlined above would still work, then have special handling to adjust the file/memsz fields after the sections have been updated.
> 
> This would either expose these temporary sections in the output file, or require some non-trivial changes to the code. And yet this does not solve the issue of gaps in `PT_NOTE` segments.

I was thinking something similar to how the ELF Header and Program Header table work, but those are pseudo-segments, not pseudo-sections. Nonetheless, I don't think it would be too hard to add a flag to the artificial section to indicate it shouldn't be included in the section header table. Again though, I'm happy for that to be deferred to a later patch, if it makes things simpler in the first one.

> > Somewhat related to the previous point, I'm concerned with how `updateSegmentData` might interact with `--remove-section` for sections in segments. The current approach of the latter is that removed section data is overwritten with null bytes, preserving the segment size. I understand that for a PT_NOTE this may not be quite so desirable, but I think we need to do something to prevent mixing the two behaviours, because what the end result should be is non-obvious. One way would be to emit an error if a note section/segment is touched by both --remove-section and --remove-note.
> 
> I suspect that there may also be some problems with using the new command together with `--add-section` and `--update-section`, as the correct order of execution is not clear. I think the simplest solution is just to reject such combinations of options, so I've updated the patch accordingly.

IIRC, we loosely model --update-section as --remove-section followed by --add-section. --add-section is deliberately done after --remove-section, to allow for a section to be replaced (and what would be the point of removing a section that was just added after all). I guess --remove-note would be similar to --update-section in that it's essentially another way of changing the contents of a note section. As such, it would come after --remove-section operations, but it's more ambiguous as to what order it applies to with regards to the other. I don't think --add-section + --remove-note interact, so I guess it doesn't matter there. --update-section + --remove-note is mostly nonsensical, where the same section is touched.

One possible exception to the above is adding/updating a note section, but wanting to remove a note from the added/modified section, because the input data wasn't quite in the right format. I don't see a use-case for --remove-note coming before the operations, but this would indicate it should come after, if it could be made to work. I also don't think the use-case is that strong, so an error is equally fine by me.

I'm not going to look at the details today - it'll have to wait until the new year when I'm back in the (home-)office.

igorkudrin committed:
Remove handling for segments

igorkudrin commented:
> Yeah, you'd need to compress the sections together and then shrink the segment. I'm not sure if there's existing code to modify a segment in this way yet.

And that exactly what my code for updating segments did. There was now existing code for anything similar. Current code expects sections to stay in the same places within segments.

> > However, the existing code is suitable for sections that are not associated with segments. I've updated the patch, so sections in relocatable objects are also supported now.
> 
> I'm happy with segment support being added later.

OK, I've removed the support for segments for now to simplify the patch. I'll create a separate review for that after this patch is landed.

> > This would either expose these temporary sections in the output file, or require some non-trivial changes to the code. And yet this does not solve the issue of gaps in `PT_NOTE` segments.
> 
> I was thinking something similar to how the ELF Header and Program Header table work, but those are pseudo-segments, not pseudo-sections. Nonetheless, I don't think it would be too hard to add a flag to the artificial section to indicate it shouldn't be included in the section header table. Again though, I'm happy for that to be deferred to a later patch, if it makes things simpler in the first one.

If we add the flag to mark a section as artificial, we will have to check it in many places, such as `ELFWriter<ELFT>::writeEhdr()`, where the size of the collection is used to calculate the value of the `e_shnum` field, or any code that scans the sections. Perhaps it would be easier to use a separate collection for them, in which case only the code that handles such sections would need to be updated.

> > I suspect that there may also be some problems with using the new command together with `--add-section` and `--update-section`, as the correct order of execution is not clear. I think the simplest solution is just to reject such combinations of options, so I've updated the patch accordingly.
> 
> IIRC, we loosely model --update-section as --remove-section followed by --add-section. --add-section is deliberately done after --remove-section, to allow for a section to be replaced (and what would be the point of removing a section that was just added after all). I guess --remove-note would be similar to --update-section in that it's essentially another way of changing the contents of a note section. As such, it would come after --remove-section operations, but it's more ambiguous as to what order it applies to with regards to the other. I don't think --add-section + --remove-note interact, so I guess it doesn't matter there. --update-section + --remove-note is mostly nonsensical, where the same section is touched.
> 
> One possible exception to the above is adding/updating a note section, but wanting to remove a note from the added/modified section, because the input data wasn't quite in the right format. I don't see a use-case for --remove-note coming before the operations, but this would indicate it should come after, if it could be made to work. I also don't think the use-case is that strong, so an error is equally fine by me.

One can argue that the intention might be to remove some unwanted notes and add new ones using these commands... or to make sure that the requested notes do not appear in the produced output, even if they are in the section added by these commands. So, to avoid speculation and overthinking, it is easier to reject such combinations.

> I'm not going to look at the details today - it'll have to wait until the new year when I'm back in the (home-)office.

Thanks!

igorkudrin commented:
Ping

jh7370 commented:
> Ping

On my TODO list. I plan on looking at this tomorrow, if I can.

jh7370 reviewed (commented):
Code mostly looks fine. I need to go over the tests again, but have run out of time for today.

jh7370 commented on llvm/test/tools/llvm-objcopy/ELF/remove-note.test:
```diff
@@ -0,0 +1,135 @@
+# ERR-INVNUM2: error: bad note type_id for --remove-note: 'Notanumber'
+
+# RUN: yaml2obj -D ALIGN=8 -D ELFCLASS=64 -D ENDIANNESS=LSB %s -o %t8.64.lsb
+# RUN: llvm-objcopy --remove-note=0x01 --remove-note=DUMMY/2 --remove-note=CORE/0x03 %t8.64.lsb %t8.64.lsb.o
```
I think you need a case that shows that `--remove-note=CORE/0x01` doesn't remove the other 0x01 instance to show that the name matching actually has an impact.

igorkudrin replied:
`--remove-note=1` checks that all notes of type `1` are removed. I've added a `LINUX/3` note to check your scenario.

jh7370 commented on llvm/lib/ObjCopy/ELF/ELFObjcopy.cpp:
```diff
@@ -609,6 +609,97 @@ static void addSymbol(Object &Obj, const NewSymbolInfo &SymInfo,
+                        ArrayRef<RemoveNoteInfo> NotesToRemove) {
+  for (auto &Sec : Obj.sections()) {
+    // TODO: Support note sections in segments
+    if (Sec.Type != SHT_NOTE || Sec.ParentSegment || !Sec.hasContents())
+      continue;
```
Perhaps worth issuing an error/warning if attempting to remove a note from a segment, to indicate it's not currently supported?

jh7370 replied:
Thanks for the update. I skimmed it and am wondering whether the callback should be part of the `Config` variable, to reduce the number of parameters being passed around. Also, in common with similar callbacks I've done elsewhere (looking at the debug line code in particular), perhaps the callback should take an `Error`, to allow for the richer information that that can contain to be passed back and checked. Will check the rest of the updates next week.

igorkudrin replied:
Okay, this will simplify the code. However, I'm not sure where is the best place to set up this callback.

jh7370 replied:
I think what you've done for now is good enough. I don't have a better suggestion at this point.

igorkudrin committed:
Remove DeletedRange.NewPos

igorkudrin committed:
"auto &RemRange" -> "const DeletedRange &RemRange"

igorkudrin committed:
Remove a blank line at the end of an namespace

igorkudrin committed:
Print a warning for note segments and note sections in segments

igorkudrin committed:
Extend the test

igorkudrin reviewed (commented):
Thanks for the review! Please have a look at the updates.

igorkudrin committed:
Revert "Print a warning for note segments and note sections in segments"

This reverts commit 56950965c937008b60c17b434f2307e8dc4e78c6.

igorkudrin committed:
removeNote() -> removeNotes()

igorkudrin committed:
Print a warning for note segments and note sections in segments v2

igorkudrin committed:
std::vector<SecPtr>::iterator -> SecPtr&

igorkudrin commented on llvm/lib/ObjCopy/ELF/ELFObject.cpp:
```diff
@@ -2154,25 +2154,20 @@ ELFWriter<ELFT>::ELFWriter(Object &Obj, raw_ostream &Buf, bool WSH,
-    return createStringError(errc::invalid_argument, "section '%s' not found",
-                             Name.str().c_str());
-
+Error Object::updateSectionData(std::vector<SecPtr>::iterator It,
```
On the other hand, we can pass `SecPtr&` instead of an iterator, which simplifies the code, and still allows the element of the container to be updated. Fixed, thanks.

igorkudrin committed:
Move tests from 'remove-note-unsupported.test' to 'remove-note.test'

jh7370 reviewed (approved):
LGTM.
